spec/ysh-funcs.test.sh

```diff
@@ -46,13 +46,71 @@ proc t() { return (0) }
 ## STDOUT:
 ## END
 
-#### Redefining functions is not allowed
+#### Redefining functions is not allowed (with shopt -u redefine_proc_func)
+shopt -u redefine_proc_func
 func f() { return (0) }
 func f() { return (1) }
 ## status: 1
 ## STDOUT:
 ## END
 
+#### Redefining functions is allowed (with shopt -s redefine_proc_func)
+shopt -s redefine_proc_func
+func f() { return (0) }
+func f() { return (1) }
+## status: 0
+## STDOUT:
+## END
+
+#### Functions cannot redefine readonly vars (even with shopt -s redefine_proc_func)
+shopt -s redefine_proc_func
+const f = 0
+func f() { return (1) }
+## status: 1
+## STDOUT:
+## END
+
+#### Functions can redefine non-readonly vars
+var f = 0
+func f() { return (1) }
+## status: 0
+## STDOUT:
+## END
+
+#### Vars cannot redefine functions (even with shopt -s redefine_proc_func)
+shopt -s redefine_proc_func
+func f() { return (1) }
+const f = 0
```

Comment:
Hmm this is interesting

Usually we try to survey Python/JavaScript for these kind of semantics

- Python doesn't have const at all
- Let's see what JS does

```
function f() { return 6 }


console.log(f)
VM175:1 ƒ f() { return 6 }


f=42
42

console.log(f)
VM194:1 42
```

OK it does let you

Reply:
Not if you use the `const` qualifier. You are just mutating `f` (I suppose we should have some tests over `setvar` on a func name). JS will complain if you try to *redeclare* `f`:

```
>> function f() { return 6 }
undefined
>> const f = 42
Uncaught SyntaxError: redeclaration of var f
    <anonymous> debugger eval code:1
debugger eval code:1:1
```

But I would argue that `func f` should behave like `const f = <...>`, with the exception of `redefined_proc_func` which adds the ergonomics of being able to write `func f` multiple times in a REPL session. So just as we cannot:

```
const f = 42
const f = 42  # error
```

We shouldn't be able to do the same in either order of `const f`, `func f`.

---

Maybe we want to change `redefine_proc_func` to `redeclare`? Then, in a REPL, we would let users redeclare variables (with `const`/`var`), funcs and procs.